packages/block-library/src/navigation/index.php

```diff
@@ -67,6 +67,521 @@ function block_core_navigation_sort_menu_items_by_parent_id( $menu_items ) {
+	 */
+	private static function does_navigation_have_submenus( $inner_blocks ) {
+		$has_submenus = false;
+		foreach ( $inner_blocks as $inner_block ) {
+			$inner_block_content = $inner_block->render();
+			$p                   = new WP_HTML_Tag_Processor( $inner_block_content );
+			if ( $p->next_tag(
+				array(
+					'name'       => 'LI',
+					'class_name' => 'has-child',
+				)
+			) ) {
+				$has_submenus = true;
+			}
+		}
+		return $has_submenus;
+	}
+
+	private static function should_load_view_script( $attributes, $inner_blocks ) {
+		$has_submenus = WP_Navigation_Block::does_navigation_have_submenus( $inner_blocks );
+		$is_responsive_menu = WP_Navigation_Block::is_responsive_navigation( $attributes );
+		return ( $has_submenus && ( $attributes['openSubmenusOnClick'] || $attributes['showSubmenuIcon'] ) ) || $is_responsive_menu;
+	}
+
+	/**
+	 * Returns the html for the inner blocks of the navigation block.
+	 */
+	private static function get_inner_blocks_html( $inner_blocks, $attributes ) {
+		$has_submenus            = WP_Navigation_Block::does_navigation_have_submenus( $inner_blocks );
+		$should_load_view_script = WP_Navigation_Block::should_load_view_script( $attributes, $inner_blocks );
+
+		$list_item_nav_blocks = array(
+			'core/navigation-link',
+			'core/home-link',
+			'core/site-title',
+			'core/site-logo',
+			'core/navigation-submenu',
+		);
+
+		$needs_list_item_wrapper = array(
+			'core/site-title',
+			'core/site-logo',
+		);
+
+		$style = WP_Navigation_Block::get_styles( $attributes );
+		$class = WP_Navigation_Block::get_classes( $attributes );
+		$container_attributes = get_block_wrapper_attributes(
+			array(
+				'class' => 'wp-block-navigation__container ' . $class,
+				'style' => $style,
+			)
+		);
+
+		$inner_blocks_html = '';
+		$is_list_open      = false;
+
+		foreach ( $inner_blocks as $inner_block ) {
+			$is_list_item = in_array( $inner_block->name, $list_item_nav_blocks, true );
+
+			if ( $is_list_item && ! $is_list_open ) {
+				$is_list_open       = true;
+				$inner_blocks_html .= sprintf(
+					'<ul %1$s>',
+					$container_attributes
+				);
+			}
+
+			if ( ! $is_list_item && $is_list_open ) {
+				$is_list_open       = false;
+				$inner_blocks_html .= '</ul>';
+			}
+
+			$inner_block_content = $inner_block->render();
+			if ( ! empty( $inner_block_content ) ) {
+				if ( in_array( $inner_block->name, $needs_list_item_wrapper, true ) ) {
+					$inner_blocks_html .= '<li class="wp-block-navigation-item">' . $inner_block_content . '</li>';
+				} else {
+					$inner_blocks_html .= $inner_block_content;
+				}
+			}
+		}
+
+		if ( $is_list_open ) {
+			$inner_blocks_html .= '</ul>';
+		}
+
+		// Add directives to the submenu if needed.
+		if ( $has_submenus && $should_load_view_script ) {
+			$w                 = new WP_HTML_Tag_Processor( $inner_blocks_html );
+			$inner_blocks_html = block_core_navigation_add_directives_to_submenu( $w, $attributes );
```

Comment:
```suggestion
			$inner_blocks_html = block_core_navigation_add_directives_to_submenu( $w, $attributes );
```

Is this function in the submenu block? If so we should probably comment by way of explanation.

Reply:
No it's in this file.

Reply:
I'm starting to wonder if we'll need a follow up PR to move all directives stuff to its own class.